### ISSUE TITLE
S3 backend on nonAWS implementation (OCI) not working in 1.6.3

### ISSUE THREAD

**Issue by 12345ieee** [hashicorp/terraform#34053]
### Terraform Version

```shell
Terraform v1.6.1
on linux_amd64
+ provider registry.terraform.io/hashicorp/null v3.2.1
+ provider registry.terraform.io/hashicorp/oci v5.9.0
+ provider registry.terraform.io/oracle/oci v5.1.0
```

### Terraform Configuration Files

Config that worked in v1.5.x:
```tf
  backend "s3" {
    bucket                      = "<BUCKET>"
    key                         = "<FILENAME>"
    region                      = "eu-frankfurt-1"
    endpoint                    = "https://<ID>.compat.objectstorage.eu-frankfurt-1.oraclecloud.com"
    shared_credentials_file     = "PATH/TO/FILE"
    skip_region_validation      = true
    skip_credentials_validation = true
    skip_metadata_api_check     = true
    force_path_style            = true
  }
```

my attempt to port it to v1.6.1, guided by the init warnings, but still failing, `skip_requesting_account_id` didn't help at all:
```tf
  backend "s3" {
    bucket                      = "<BUCKET>"
    key                         = "<FILENAME>"
    region = "eu-frankfurt-1"
    endpoints = {
      s3 = "https://<ID>.compat.objectstorage.eu-frankfurt-1.oraclecloud.com"
    }
    shared_credentials_files    = ["PATH/TO/FILE"]
    skip_region_validation      = true
    skip_credentials_validation = true
    skip_metadata_api_check     = true
    skip_requesting_account_id  = true
    use_path_style              = true
  }
```

### Debug Output

See below

### Expected Behavior

For my init to go through, like it did in tf 1.5.x with the old config.

In all honesty I'd have preferred not to need an init at all, but I can live with this.

### Actual Behavior

The relevant log part is thankfully short:
`$ TF_LOG=trace terraform init -reconfigure` 
```
[...]
2023-10-11T01:43:32.025+0200 [TRACE] backend-s3.aws-base: Resolving AWS configuration: tf_backend.operation=Configure tf_backend.req_id=4951b714-0978-3879-4889-51e8b4093226 tf_backend.s3.bucket=terraform-state tf_backend.s3.path=[REDACTED]
2023-10-11T01:43:32.025+0200 [DEBUG] backend-s3.aws-base: Resolving credentials provider: tf_backend.operation=Configure tf_backend.req_id=4951b714-0978-3879-4889-51e8b4093226 tf_backend.s3.bucket=terraform-state tf_backend.s3.path=[REDACTED]
2023-10-11T01:43:32.025+0200 [TRACE] backend-s3.aws-base: Building default HTTP client: tf_backend.operation=Configure tf_backend.req_id=4951b714-0978-3879-4889-51e8b4093226 tf_backend.s3.bucket=terraform-state tf_backend.s3.path=[REDACTED]
2023-10-11T01:43:32.025+0200 [DEBUG] backend-s3.aws-base: Loading configuration: tf_backend.operation=Configure tf_backend.req_id=4951b714-0978-3879-4889-51e8b4093226 tf_backend.s3.bucket=terraform-state tf_backend.s3.path=[REDACTED]
2023-10-11T01:43:32.025+0200 [DEBUG] backend-s3.aws-base: Retrieving credentials: tf_backend.operation=Configure tf_backend.req_id=4951b714-0978-3879-4889-51e8b4093226 tf_backend.s3.bucket=terraform-state tf_backend.s3.path=[REDACTED]
╷
│ Error: No valid credential sources found
│ 
│ Please see https://www.terraform.io/docs/language/settings/backends/s3.html
│ for more information about providing credentials.
│ 
│ Error: failed to refresh cached credentials, no EC2 IMDS role found, operation error ec2imds: GetMetadata, access disabled to EC2 IMDS via client option, or
│ "AWS_EC2_METADATA_DISABLED" environment variable
│ 
╵
```

### Steps to Reproduce

`$ terraform init -reconfigure`

### Additional Context

The documentation at https://developer.hashicorp.com/terraform/language/settings/backends/s3 has not been updated, that didn't help.

### References

- #34022 
- #34002

**Comment by hegerdes** [comment]
I can get it to initialize but apply and plan fails to persist data to the backend.

For testing i used hard-coded secrets:
```hcl
  backend "s3" {
    bucket                      = "<BUCKET>"
    key                         = "<FILENAME>"
    region = "eu-frankfurt-1"
    endpoints = {
      s3 = "https://my-dell-s3.com"
    }
    access_key = "<MY_KEY>"
    secret_key = "<MY_SECRET>"
    skip_region_validation      = true
    skip_credentials_validation = true
    skip_metadata_api_check     = true
    skip_requesting_account_id  = true
    use_path_style              = true
  }
```

As I said init works but plan and destroy fails with:
```
Error: Failed to save state

Error saving state: failed to upload state: operation error S3: PutObject, https response error StatusCode: 400, RequestID: 0a3d0687:189c57fbda4:59f00:c15, HostID:
081a966b5d23d91a6a72d9bcc2f38088d47ca56d8e6c7ee9b2c0055d044ae504, api error XAmzContentSHA256Mismatch: The Content-SHA256 you specified did not match what we received

Error: Failed to persist state to backend

The error shown above has prevented Terraform from writing the updated state to the configured backend. To allow for recovery, the state has been written to the file
"errored.tfstate" in the current working directory.

Running "terraform apply" again at this point will create a forked state, making it harder to recover.

To retry writing this state, use the following command:
    terraform state push errored.tfstate
```

Honestly all these S3 related errors do not shine a good light on hashicorps testing processes. These are not some abstruse functions that do not work, its basic functionality.

**Comment by 12345ieee** [comment]
I didn't try with hardcoded credentials because it's an hard requirement for me to have them in a file, but now I'm even less inclined to do more tests.

thanks for the warning @hegerdes

**Comment by mmeier86** [comment]
@hegerdes For what it's worth, I'm seeing the same error using Ceph's S3 compatible storage. Because @12345ieee seems to have a slightly different problem, I opened a separate bug at #34086.

**Comment by holms** [comment]
Same issue with digital ocean spaces:

```hcl
terraform {
  backend "s3" {
    skip_region_validation      = true
    skip_credentials_validation = true
    skip_metadata_api_check     = true
    skip_requesting_account_id  = true
    use_path_style              = true
    endpoints = {
      s3 = "https://fra1.digitaloceanspaces.com"
    }
    region                      = "fra1" // needed
    bucket                      = "thorauto-terraform"
    key                         = "terraform.tfstate"
   }
}
```

```bash
│ Error: Failed to save state
│ 
│ Error saving state: failed to upload state: operation error S3: PutObject, https response error StatusCode: 400, RequestID: tx00000b96657e5f692e5d8-00652f2405-bc9d9e44-fra1b, HostID: bc9d9e44-fra1b-fra1-zg02, api error XAmzContentSHA256Mismatch: UnknownError
╵

```

**Comment by zeyad-sitech** [comment]
Same issue here...

**Comment by jar-b** [comment]
Unfortunately we are unable to test changes to the S3 backend against the various S3-compatible storage providers, though it is never our intention to break existing workflows. I would recommend raising this with the upstream providers as the behavior has been confirmed to work correctly with Amazon S3. As part of the S3 Backend maintenance we did move from using [v1](https://github.com/aws/aws-sdk-go) of the AWS Go SDK to [v2](https://github.com/aws/aws-sdk-go-v2), specifically adopting the [S3 Manager](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/feature/s3/manager) feature which improves performance on large state files. It's possible that these providers have not implemented this functionality. We may be able to help resolve these issues with more information from those provider teams, but at this time cannot commit to investigating further.

**Comment by hegerdes** [comment]
@jar-b Thanks for the statement but this is not a great perspective going forward. It is totally clear that Hashicorp can't test against all of the existing S3 implementations but is not like there are problems with just one provider. Digital Ocean and Ceph are not some small vendors. We are also having problems with a system from Dell. Your comment implies that it wasn't testet against ANY other provider than aws, which (as you said) was not documented. The documentation just said that S3 works if the required permissions (put, list, etc) are fulfilled. It also looks like that there is not much interest in fixing this problem when pointing to the various S3 vendors - maybe it's because Hashicorp wants to push it's own cloud backend?
Version 1.6.x is just not usable for a lot of people. This kind of stuff and behavior is the reason why so many people want to see openTofu to succeed.

**Comment by crw** [comment]
@hegerdes Thanks for the feedback. For vendors which offer "S3-compatible" services, the burden of compatibility falls on those vendors. HashiCorp is supporting a backend for the AWS S3 service, and is leveraging the Golang AWS SDK to do so. As AWS updates its SDK, other competing services may fall behind on compatibility for some amount of time. We plan to update the S3 backend documentation to make this nuance of using the S3 backend more explicit. 

I apologize for any frustration this may cause. Thanks again for your continued feedback on this issue!

**Comment by gdavison** [comment]
Hello everyone. @12345ieee's initial report was a failure due to failing authentication:

> Error: No valid credential sources found

A number of other people have reported different errors with their use of "S3-compatible" services, all related to the error `XAmzContentSHA256Mismatch`. This is related to the `ChecksumAlgorithm` S3 API parameter that is required in some cases for S3 and recommended in others. It appears that most "S3-compatible" services either don't support the parameter or don't support setting certain values.

Typically, we ask for separate issues for separate problems. In this case, we already have several issues related to `XAmzContentSHA256Mismatch`, so that is not required.

**Comment by gdavison** [comment]
@12345ieee, can you please share your shared credentials file (with sensitive values blanked out)

**Comment by 12345ieee** [comment]
Sure @gdavison , here you go:
```
[default]
aws_access_key_id = [REDACTED]
aws_secret_access_key = [REDACTED]
```

**Comment by rp-jasonp** [comment]
Seeing slightly different errors, following thread.

Hashicorp: Please revert changes to s3 backend provider, create s3_v2 provider or some such solution going forward.

We may look to accelerate moving our states to Artifactory.

```
terraform {
  backend "s3" {
    bucket                      = "<BUCKET>"
    key                         = "<STATE KEY>"
    region                      = "us-phoenix-1"
    endpoint                    = "https://<ID>.compat.objectstorage.us-phoenix-1.oraclecloud.com"
    skip_region_validation      = true
    skip_credentials_validation = true
    skip_metadata_api_check     = true
    force_path_style            = true
  }
}
```


```
 Error: Retrieving AWS account details: AWS account ID not previously found and failed retrieving via all available methods. See https://www.terraform.io/docs/providers/aws/index.html#skip_requesting_account_id for workaround and implications. Errors: 2 errors occurred:
 │       * retrieving caller identity from STS: operation error STS: GetCallerIdentity, https response error StatusCode: 0, RequestID: , request send failed, Post "https://sts.us-phoenix-1.amazonaws.com/": dial tcp: lookup sts.us-phoenix-1.amazonaws.com: no such host
 │       * retrieving account information via iam:ListRoles: operation error IAM: ListRoles, https response error StatusCode: 403, RequestID: <REQID from OCI> , api error InvalidClientTokenId: The security token included in the request is invalid.
```

**Comment by jar-b** [comment]
@rp-jasonp - you may also need to set the `skip_requesting_account_id` argument:

https://developer.hashicorp.com/terraform/language/settings/backends/s3#skip_requesting_account_id

**Comment by rp-jasonp** [comment]
No dice. 

`Post "https://sts.us-phoenix-1.amazonaws.com/":`

Oracle's official doc reference: https://docs.oracle.com/en-us/iaas/Content/API/SDKDocs/terraformUsingObjectStore.htm

**Comment by gdavison** [comment]
@rp-jasonp, since this is a different problem from @12345ieee's (and not related to `XAmzContentSHA256Mismatch`), please open a separate issue.

**Comment by mark-wff** [comment]
Same error here: 
```
terraform {
  backend "s3" {
    endpoints = {
      s3 = "https://fra1.digitaloceanspaces.com"
    }
    region = "fra1"
    bucket = ""
    key    = "key"
    secret_key = ""
    access_key = ""
    skip_region_validation = true
    skip_requesting_account_id = true
    skip_credentials_validation = true
    skip_metadata_api_check = true
  }
```

```
╷
│ Error: Error saving current state
│ 
│ Error: failed to upload state: operation error S3: PutObject, https response error StatusCode: 400, RequestID: tx0000082ece50911b02170-00653a0845-bc9d8cdc-fra1b, HostID: bc9d8cdc-fra1b-fra1-zg02, api error XAmzContentSHA256Mismatch: UnknownError
│ 
│ Terraform encountered an error attempting to save the state before cancelling the current operation. Once the operation is complete another attempt will be made to save the final state.
```

**Comment by frittentheke** [comment]
I recently ran into issues with mountpoint-s3 not working with Ceph RADOSGW with the `XAmzContentSHA256Mismatch` error being thrown. I know mountpoint-s3 is written in Rust ... but still it's a very current S3 library and likely AWS is doing something "new" here. Issue lives at: https://tracker.ceph.com/issues/63153

**Comment by vineelachavali** [comment]
I do use OCI and its failing with same latest error , can some one please look into it?
```│ Warning: Deprecated Parameter
│
│   on variables.tf line 44, in terraform:
│   44:     force_path_style            = true
│
│ The parameter "force_path_style" is deprecated. Use parameter
│ "use_path_style" instead.
╵

╷
│ Error: Retrieving AWS account details: AWS account ID not previously found and failed retrieving via all available methods. See https://www.terraform.io/docs/providers/aws/index.html#skip_requesting_account_id for workaround and implications. Errors: 2 errors occurred:
│       * retrieving caller identity from STS: operation error STS: GetCallerIdentity, https response error StatusCode: 0, RequestID: , request send failed, Post "https://sts.ap-tokyo-1.amazonaws.com/": dial tcp: lookup sts.ap-tokyo-1.amazonaws.com: no such host
│       * retrieving account information via iam:ListRoles: operation error IAM: ListRoles, https response error StatusCode: 403, RequestID: 73b70f40-d5ac-4a65-9d38-3c89ecaf96d7, api error InvalidClientTokenId: The security token included in the request is invalid.

**Comment by yordis** [comment]
```

**Comment by yordis** [comment]
Any idea of what to do here? My setup just stopped working and I can't tell what do anymore. I seriously would appreciate any guidelines if you got it to work with Digital Ocean.

```
╷
Initializing modules...
│ Error: No valid credential sources found
│ 
│ Please see https://www.terraform.io/docs/language/settings/backends/s3.html
│ for more information about providing credentials.
│ 
│ Error: failed to refresh cached credentials, no EC2 IMDS role found,
│ operation error ec2imds: GetMetadata, access disabled to EC2 IMDS via
│ client option, or "AWS_EC2_METADATA_DISABLED" environment variable
│ 
╵
```

```hcl
terraform {
  backend "s3" {
    bucket = "[REDACTED]"
    endpoints = {
      s3 = "https://nyc3.digitaloceanspaces.com"
    }
    key                         = "[REDACTED]/terraform.tfstate"
    region                      = "us-east-1"
    skip_credentials_validation = true
    skip_metadata_api_check     = true
    skip_requesting_account_id  = true
    skip_s3_checksum            = true
  }
}
```

---

I am really disliking Hashicorp comment.


I am not even a Ops dev so I can not afford such breaking changes when I have the very minimal skills to get things done, and I can't give up with such horrific way to treat the ecosystem to say the least. Really lazy to blame others for your failures.

I do not appreciate it a bit!

**Comment by crw** [comment]
Hi @yordis, thank for your comment. Just a reminder to please follow the [Community Guidelines](https://www.hashicorp.com/community-guidelines) when posting. 

As a quick suggestion, unless you specifically needed one of the new features from 1.6, you can continue to use `1.5.7` where your old config will continue to work.

**Comment by holms** [comment]
> Any idea of what to do here? My setup just stopped working and I can't tell what do anymore. I seriously would appreciate any guidelines if you got it to work with Digital Ocean.
> 
> ```
> terraform 1.6.3
> ```
> 
> ```
> ╷
> Initializing modules...
> │ Error: No valid credential sources found
> │ 
> │ Please see https://www.terraform.io/docs/language/settings/backends/s3.html
> │ for more information about providing credentials.
> │ 
> │ Error: failed to refresh cached credentials, no EC2 IMDS role found,
> │ operation error ec2imds: GetMetadata, access disabled to EC2 IMDS via
> │ client option, or "AWS_EC2_METADATA_DISABLED" environment variable
> │ 
> ╵
> ```
> 
> ```terraform
> terraform {
>   backend "s3" {
>     bucket = "[REDACTED]"
>     endpoints = {
>       s3 = "https://nyc3.digitaloceanspaces.com"
>     }
>     key                         = "[REDACTED]/terraform.tfstate"
>     region                      = "us-east-1"
>     skip_credentials_validation = true
>     skip_metadata_api_check     = true
>     skip_requesting_account_id  = true
>     skip_s3_checksum            = true
>     skip_region_validation      = true
>   }
> }
> ```
> 
> I am really disliking Hashicorp comment.
> 
> I am not even a Ops dev so I can not afford such breaking changes when I have the very minimal skills to get things done, and I can't give up with such horrific way to treat the ecosystem to say the least. Really lazy to blame others for your failures.
> 
> I do not appreciate it a bit!

I'm also furious in here, so I've contacted digital ocean regarding how they gonna deal with this. They said they gonna try to fix api on their side so that it would work with latest terraform. I believe we need to create a ticket for them in [digitalocean/terraform-provider-digitalocean ](https://github.com/digitalocean/terraform-provider-digitalocean/) repo. Also I've asked for the update by email just now as they had internal discussion about this for the last couple of weeks.

**Comment by holms** [comment]
> Hi @yordis, thank for your comment. Just a reminder to please follow the [Community Guidelines](https://www.hashicorp.com/community-guidelines) when posting.
> 
> As a quick suggestion, unless you specifically needed one of the new features from 1.6, you can continue to use `1.5.7` where your old config will continue to work.

If people are furious, I believe something should be done about this. It's absolutely unacceptable to avoid backward compatibility in such manner. Also it's quite sad that hashicorp doesn't partner with top10 biggest cloud providers out there before doing such changes, so they'd implement support for new aws sdk.

**Comment by holms** [comment]
Ok here's update from Digital ocean:

>A change has been merged into TF that should obviate this problem: https://github.com/hashicorp/terraform/pull/34130
>The release for that change was TF v1.6.3 (according to the description), which happened on the 1st of November. I haven't >yet tried this myself, though.

**Comment by frittentheke** [comment]
On the Ceph RADOSGW side (which is the software that quite a lot of non-AWS S3 services use) there already are plans to implement the missing feature for the upcoming version: https://tracker.ceph.com/issues/63153#note-8

**Comment by sosiskin2000** [comment]
+1 for OCI, after apply
Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
╷
│ Error: Failed to save state
│ 
│ Error saving state: failed to upload state: operation error S3: PutObject, https response error StatusCode: 400, RequestID: lhr-1:cZsaCiYlaGoJrJ-pwZOI8Yqi_XIk71-YxVnQphsaqSO8Y6FXtsmsD5NxKMhYEFED, HostID: , api error InvalidArgument:
│ x-amz-content-sha256 must be UNSIGNED-PAYLOAD or a valid sha256 value.
╵
╷
│ Error: Failed to persist state to backend
│ 
│ The error shown above has prevented Terraform from writing the updated state to the configured backend. To allow for recovery, the state has been written to the file "errored.tfstate" in the current working directory.
│ 
│ Running "terraform apply" again at this point will create a forked state, making it harder to recover.
│ 
│ To retry writing this state, use the following command:
│     terraform state push errored.tfstate

**Comment by matkovskiy** [comment]
Hello. 
To fix this add skip_s3_checksum = true to terraform backend configuration. 
Configuration which works for me 
`terraform {
  backend "s3" {
    bucket = "BUCKET_NAME"
    region = "REGION"
    key = "tf.tfstate"
    skip_region_validation = true
    skip_credentials_validation = true
    skip_requesting_account_id = true
    use_path_style = true
    insecure = true
    skip_metadata_api_check = true
    skip_s3_checksum = true
    access_key = "XXXXXX"
    secret_key = "YYYYYYYYYY"
    endpoints = { s3 = "https://ID.compat.objectstorage.REGION.oraclecloud.com" }
  }
}
`

P.s. access_key  secret_key just for demo, don't use it in real environment 

P.p.s from https://github.com/hashicorp/terraform/issues/34086

**Comment by sosiskin2000** [comment]
> To fix this add skip_s3_checksum = true to terraform backend configuration.

Works for me, thank you.

**Comment by alecaugh** [comment]
[terraform version 1.6.6]
Unfortunately this workaround doesn't seem to work when wanting to provide the key via environmental variables i.e.

```
terraform {
  required_version = "1.6.6"

  required_providers {
    cloudflare = {
      source  = "cloudflare/cloudflare"
      version = "~> 4"
    }
  }
}

data "terraform_remote_state" "state" {
  backend = "s3"

  config = {
    bucket    = "terraform-state"
    key       = "project_name/terraform.tfstate"
    endpoints = { s3 = "https://xxxx.r2.cloudflarestorage.com" }
    region    = "us-east-1"

    access_key                  = "${var.CLOUDFLARE_S3_ACCESS_KEY_ID}"
    secret_key                  = "${var.CLOUDFLARE_S3_SECRET_ACCESS_KEY}"
    skip_credentials_validation = true
    skip_region_validation      = true
    skip_requesting_account_id  = true
    skip_metadata_api_check     = true
    skip_s3_checksum            = true
  }
}
```

It's returning this as an error during the `terraform plan`:

```
 Error: Unable to access object "terraform.tfstate" in S3 bucket "terraform": operation error S3: HeadObject, https response error StatusCode: 400, RequestID: , HostID: , api error BadRequest: Bad Request
│```

Has anyone else encountered this issue and found a solution?

**Comment by gantta** [comment]
Encountered this with Terraform version `v1.7.4` as well. 
```zsh
❯ terraform version
Terraform v1.7.4
on darwin_arm64
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/oracle/oci v5.30.0
```
Using OCI for remote state S3 backend and [Oracle's documentation](https://docs.oracle.com/en-us/iaas/Content/API/SDKDocs/terraformUsingObjectStore.htm#s3), I encountered the same surprising error message AFTER `terraform apply` created the resources:
```zsh
╷
│ Error: Failed to save state
│ 
│ Error saving state: failed to upload state: operation error S3: PutObject, https response error StatusCode: 400, RequestID:
│ iad-1:B9UETKCkS7JqOriMsjfljnZXgJ_Nh6nqtl3R-VgJB5zhfj6mZueR-Vm_xviWX-e1, HostID: , api error InvalidArgument: x-amz-content-sha256
│ must be UNSIGNED-PAYLOAD or a valid sha256 value.
╵
╷
│ Error: Failed to persist state to backend
│ 
│ The error shown above has prevented Terraform from writing the updated state to the configured backend. To allow for recovery, the
│ state has been written to the file "errored.tfstate" in the current working directory.
│ 
│ Running "terraform apply" again at this point will create a forked state, making it harder to recover.
│ 
│ To retry writing this state, use the following command:
│     terraform state push errored.tfstate
│ 
╵
```
Reading through the comments above, I added the `skip_s3_checksum = true`  configuration and was able to successfully reconfigure the backend and upload the errored statefile:
```zsh
❯ terraform init -reconfigure

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

Initializing provider plugins...
- Reusing previous version of hashicorp/null from the dependency lock file
- Reusing previous version of oracle/oci from the dependency lock file
- Using previously-installed hashicorp/null v3.1.0
- Using previously-installed oracle/oci v5.30.0

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

❯ terraform state push errored.tfstate
```
And I can verify the state file is now present in my OCI bucket.
```zsh
> oci os object list -ns <my-bucket-namespace> -bn <my-bucket>                                                      
{
  "data": [
    {
      "archival-state": null,
      "etag": "da5b84c7-d80f-4fb1-b68c-14352985a975",
      "md5": "PFvsKR2ny4ahwFY481W6KA==",
      "name": "tf-landing-zone.tfstate",
      "size": 5432,
      "storage-tier": "Standard",
      "time-created": "2024-03-07T07:14:39.992000+00:00",
      "time-modified": "2024-03-07T07:14:39.992000+00:00"
    }
  ],
  "prefixes": []
}
```

**Comment by frittentheke** [comment]
> On the Ceph RADOSGW side (which is the software that quite a lot of non-AWS S3 services use) there already are plans to implement the missing feature for the upcoming version: https://tracker.ceph.com/issues/63153#note-8

The missing checksum features / bugfixes were merged to Ceph master three weeks ago: https://github.com/ceph/ceph/pull/54856. It's also planned to backport them to the Reef and Quincy releases, but that is still ongoing.

**Comment by noprobsdk** [comment]
Hi

Im using TF/OCI, and I encounted this error today. Worked fine yesterday..

Error: Failed to get existing workspaces: Unable to list objects in S3 bucket "**-terraform-state-***": operation error S3: ListObjectsV2, https response error StatusCode: 403, RequestID: fra-1:pWnsaVPzVkERWDRo**************RY3eb8kjAEbaajwyCvRZFjSzksi, HostID: , api error SignatureDoesNotMatch: The secret key required to complete authentication could not be found. The region must be specified if this is not the home region for the tenancy.

My backend.tf:

terraform {
  backend "s3" {
    bucket                        = "**-terraform-state-***"
    key                           = "./terraform.tfstate"
    region                        = "eu-frankfurt-1"
    endpoints                     = {s3 = "https://frm**crv.compat.objectstorage.eu-frankfurt-1.oraclecloud.com"}
    shared_credentials_files       = ["~/.oci/aws_s3_compatible_credentials"]
    skip_region_validation        = true
    skip_credentials_validation   = true
    skip_metadata_api_check       = true
    use_path_style                = true
    skip_requesting_account_id    = true
    skip_s3_checksum              = true
  }
}

Tried everything, also old backend.tf with TF v1.5.x

All development has stopped, as we cant deploy. Oracle Support says they have several reports on this.

**Comment by AAverin** [comment]
on 1.7.4
Getting `Error: No valid credential sources found` for Digital Ocean Spaces

**Comment by patilak1** [comment]
> Hello. 
> To fix this add skip_s3_checksum = true to terraform backend configuration. 
> Configuration which works for me 
> `terraform {
>   backend "s3" {
>     bucket = "BUCKET_NAME"
>     region = "REGION"
>     key = "tf.tfstate"
>     skip_region_validation = true
>     skip_credentials_validation = true
>     skip_requesting_account_id = true
>     use_path_style = true
>     insecure = true
>     skip_metadata_api_check = true
>     skip_s3_checksum = true
>     access_key = "XXXXXX"
>     secret_key = "YYYYYYYYYY"
>     endpoints = { s3 = "https://ID.compat.objectstorage.REGION.oraclecloud.com" }
>   }
> }
> `
> 
> P.s. access_key  secret_key just for demo, don't use it in real environment 
> 
> P.p.s from https://github.com/hashicorp/terraform/issues/34086

How to refer this access key and secret key from GitHub ontime runner file i.e /.AWS/credentials file here ... It tried directly by echo keys but it didn't work.. can pls suggest what could be the configuration in main.tf and in workflow for the backend

**Comment by patilak1** [comment]
> > Hello.
> > To fix this add skip_s3_checksum = true to terraform backend configuration.
> > Configuration which works for me
> > `terraform { backend "s3" { bucket = "BUCKET_NAME" region = "REGION" key = "tf.tfstate" skip_region_validation = true skip_credentials_validation = true skip_requesting_account_id = true use_path_style = true insecure = true skip_metadata_api_check = true skip_s3_checksum = true access_key = "XXXXXX" secret_key = "YYYYYYYYYY" endpoints = { s3 = "https://ID.compat.objectstorage.REGION.oraclecloud.com" } } } `
> > P.s. access_key  secret_key just for demo, don't use it in real environment
> > P.p.s from #34086
> 
> How to refer this access key and secret key from GitHub ontime runner file i.e /.AWS/credentials file here ... It tried directly by echo keys but it didn't work.. can pls suggest what could be the configuration in main.tf and in workflow for the backend

terraform {
  backend "s3" {
    bucket                      = "xxx"
    key                         = "xx"
    region                      = "xxxx"
    skip_region_validation      = true
    skip_credentials_validation = true
    skip_metadata_api_check     = true
    skip_requesting_account_id  = true
    use_path_style              = true
    skip_s3_checksum            = true
    access_key                  = "~/.aws/credentials"
    secret_key                  = "~/.aws/credentials"
    endpoints                   = { s3 = "xxxx" }
  }
}
+++++
In github workflow
- name: 'Configure default profile'
        shell: bash
        run: |
          mkdir ~/.aws
          echo "[default]">>~/.aws/credentials
          echo access_key="$AWS_ACCESS_KEY_ID" >> ~/.aws/credentials
          echo secret_key="$AWS_SECRET_ACCESS_KEY" >> ~/.aws/credentials
          chmod ugo+rwx ~/.aws/credentials
Getting Error in Terraform init
Error: Failed to get existing workspaces: Unable to list objects in S3 bucket "***": operation error S3: ListObjectsV2, https response error StatusCode: 400, RequestID: arn-1:1xxxxxx, HostID: , api error AuthorizationHeaderMalformed: Unable to parse Authorization header AWS4-HMAC-SHA256 Credential=~/.aws/credentials/20240311/***/s3/aws4_request, SignedHeaders=accept-encoding;amz-sdk-invocation-id;host;x-amz-content-sha256;x-amz-date, Signature=4xxxxx

**Comment by noprobsdk** [comment]
Hi, OCI must have something, as it works now :) Both TF's on 1.5.x and 1.6.x

**Comment by patilak1** [comment]
> Hi, OCI must have something, as it works now :) Both TF's on 1.5.x and 1.6.x

could you possibly share me the reference code for configuration

**Comment by 12345ieee** [comment]
I can confirm this style works now, on 1.7.4:
```tf
  backend "s3" {
    bucket = "[REDACTED]"
    key    = "[REDACTED]"
    region = "eu-frankfurt-1"
    endpoints = {
      s3 = "https://[REDACTED].compat.objectstorage.eu-frankfurt-1.oraclecloud.com"
    }
    profile = "[REDACTED]"
    skip_credentials_validation = true
    skip_metadata_api_check     = true
    skip_region_validation      = true
    skip_requesting_account_id  = true
    skip_s3_checksum            = true
    use_path_style              = true
  }
```
using the default credentials file, but no env vars.

**Comment by vadymkiz** [comment]
> I have the same problem again.. Wrote to Oracle support
> 
> Error: error loading state: SignatureDoesNotMatch: The secret key required to complete authentication could not be found. The region must be specified if this is not the home region for the tenancy. │ status code: 403, request id: fra-1:wfAdQHeRFTrVkHzIt_btw4r-CfDKvQEdTuv6w3BS4dbfqOXPg_2-DPgZ46ri3yIG, host id:

me too, it appears absolutely randomly
even s3 cli doesn't work
